google/cloud/storage/internal/access_control_common.h

```diff
@@ -151,6 +140,25 @@ class AccessControlCommon {
   std::string self_link_;
 };
 
+inline bool operator==(AccessControlCommon const& lhs,
+                       AccessControlCommon const& rhs) {
+  // Start with id, bucket, etag because they should fail early, then
+  // alphabetical for readability.
+  return lhs.id() == rhs.id() && lhs.bucket() == rhs.bucket() &&
+         lhs.etag() == rhs.etag() && lhs.domain() == rhs.domain() &&
+         lhs.email() == rhs.email() && lhs.entity() == rhs.entity() &&
+         lhs.entity_id() == rhs.entity_id() && lhs.kind() == rhs.kind() &&
+         lhs.has_project_team() == rhs.has_project_team() &&
+         (!lhs.has_project_team() ||
```

Comment:
I am confused by these two lines. Would `lhs.has_project_team() && rhs.has_project_team()` instead of `lhs.has_project_team() == rhs.has_project_team()` be more readable for the intent here? Also, I think this will evaluate to true if `lhs.has_project_team` is `false`. I'm not sure if that's what you want.

Reply:
Maybe it's a little clever but it also seems like a pretty well-established idiom for writing this sort of thing without using ifs (although [optional's operator==](https://github.com/googleapis/google-cloud-cpp/blob/440439dcd72f362ff5eccfea34790a36b32a8523/google/cloud/optional.h#L235) takes that approach). The logic is the same, for them to be equal we need both:

1. lhs and rhs must both either: have a value, or not have a value (hence == rather than &&)
2. if they both have a value, those values must be equal

clause 2 can be restated as "either they both **don't** have values, or those values are equal" - except since clause 1 already established equality of has_XXX() we only need to check one.

Reply:
Thanks for the explanation. 

> lhs and rhs must both either: have a value, or not have a value (hence == rather than &&)

If they don't have a value, then we **don't** want to run the following check (otherwise we are deferencing a NULL pointer). Correct me if I'm wrong.

> if they both have a value, those values must be equal

What I'm concerned about is `if (!flag || a == b)`. As I understand it, if flag is `false`, the entire statement is `true`.

Reply:
OK disregard, your comment. I took a look at the implementation for optional and your approach LGTM.